src/number_theory/pell_general.lean

```diff
@@ -0,0 +1,120 @@
+/-
+Copyright (c) 2023 Michael Stoll. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Michael Geißer, Michael Stoll
+-/
+
+import tactic.qify
+import data.zmod.basic
+import number_theory.diophantine_approximation
+
+/-!
+# Pell's Equation
+
+We prove the following
+
+**Theorem.** Let $d$ be a positive integer that is not a square. Then the equation
+$x^2 - d y^2 = 1$ has a nontrivial (i.e., with $y \ne 0$) solution in integers.
+
+See `pell.ex_nontriv_sol`.
+
+This is the beginning of a development that aims at providing all of the essential theory
+of Pell's Equation for general $d$ (as opposed to the contents of `number_theory.pell`,
+which is specific to the case $d = a^2 - 1$ for some $a > 1$).
+
+## References
+
+* [K. Ireland, M. Rosen, *A classical introduction to modern number theory*
+   (Section 17.5)][IrelandRosen1990]
+
+## Tags
+
+Pell's equation
+-/
+
+namespace pell
+
+section existence
+
+open set real
+
+/-- If `d` is a positive integer that is not a square, then there is a nontrivial solution
+to the Pell equation `x^2 - d*y^2 = 1`. -/
+theorem ex_nontriv_sol {d : ℤ} (h₀ : 0 < d) (hd : ¬ is_square d) :
```

Comment:
Does it make sense to use `d : nat` instead? `d = 0` is already forbidden by `hd`.

Reply:
Maybe there's no real point

Reply:
In fact, the first version of this was for natural `d`, but I think for applications integral `d` is probably more useful. The statement of the assumption as "`d` is positive and non-square" is standard in this context.
(Of course, with `d : nat`, one could dispense with `h₀` (at the cost of having to deduce that `d` is nonzero from `hd`), but this version would be more painful to use with a parameter that is of type `int`, whereas a version for `d : nat` can be easily deduced from the integral version.)